### PR TITLE
Blockchain database: Refactor ProofOfWork

### PR DESCRIPTION
**Pull request by CjS77**
In preparation for a backend-agnostic BC-DB, we refactor the proof of work slightly.

The concept of a target and difficulty is laid out, and a demo PoW based on Blake2b is provided
that's slightly more functional than the existing `MockPoW`, including a simple
mining algorithm.

The current PoW modules remains in place for now -- an upcoming PR will replace it.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:

* [ ] I'm merging against the `development` branch
* [ ] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/core/src/proof_of_work/blake_pow.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+///
+/// The proof of work difficulty is given by `H256(H512(header || nonce))` where Hnnn is the Blake2b digest of length
+/// `nnn` bits.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct BlakePow;
+
+impl BlakePow {
+    /// A simple miner. It starts at nonce = 0 and iterates until it finds a header hash that meets the desired target
+    pub fn mine(target_difficulty: Difficulty, header: &BlockHeader) -> u64 {
+        let mut nonce = 2u64;
```
Comment (philipr-za) [inline base_layer/core/src/proof_of_work/blake_pow.rs]: Nit: out of sync with the comment.

File: base_layer/core/src/proof_of_work/difficulty.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+impl Default for Difficulty {
+    fn default() -> Self {
+        Difficulty(0)
+    }
+}
+
+// You can only add or subtract Difficulty from Difficulty
+newtype_ops! { [Difficulty] {add sub} {:=} Self Self }
```
Comment (philipr-za) [inline base_layer/core/src/proof_of_work/difficulty.rs]: what sorcery is this!?

Comment (CjS77) [reply base_layer/core/src/proof_of_work/difficulty.rs]: Magic supplied by the nifty newtype_ops crate :)

File: base_layer/core/src/proof_of_work/blake_pow.rs (hunk 1 above)
Comment (CjS77) [reply base_layer/core/src/proof_of_work/blake_pow.rs]: Good catch

File: base_layer/core/src/proof_of_work/pow.rs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+use crate::{
+    blocks::blockheader::BlockHeader,
+    proof_of_work::{difficulty::Difficulty, error::PowError},
+};
+use tari_utilities::{ByteArray, Hashable};
+
+/// `WorkProof` is a trait that captures common functionality for different proof of work algorithms.
```
Comment (mikethetike) [inline base_layer/core/src/proof_of_work/pow.rs]:
```suggestion
///  `ProofOfWork` is a trait that captures common functionality for different proof of work algorithms.
```

**Pull request merged**